### ISSUE TITLE
FR: support DNS names that refer to "any node with tag T"

### ISSUE THREAD

**Issue by josharian** [tailscale/tailscale#4324]
### What are you trying to do?

I'm spinning up tagged ephemeral Tailscale nodes on fly.io to serve our internal dashboard.

Every time I deploy, new nodes get created and the old nodes get killed.

As a result, I have no user-friendly way to reach the dashboard.

I can't use IPs because those change every time an ephemeral node gets created. I can't use hostname-based DNS because the lifetimes of the nodes overlap, so the entries end up with numeric suffixes.

### How should we solve this?

I'd like to be able to have a DNS name that maps to "any currently-online node with tag T". Maybe something like this in the ACL:

```
 "hosts": {
    "dashboard": ["tag:dashboard:*"],
  },
```

There are lots of potential semantics for how to resolve the DNS name (most recently seen, load balance, lowest ping latency / first to respond to peer request, random). I don't really care which the semantics are.

I'd also be happy with other ways to fix this, as long as they're low effort + low maintenance.

### What is the impact of not solving this?

I will either (1) hack together a script to update the ACL file after every deployment; (2) play subnet router games, as suggested by @maisem, or mostly realistically (3) copy/paste IP addresses until my brain explodes and then go do 1 or 2.

### Anything else?

Related: https://github.com/tailscale/tailscale/issues/465

Somewhat related: https://github.com/tailscale/tailscale/issues/1543

cc @maisem @crawshaw @mayakacz

**Comment by maisem** [comment]
>play subnet router games, as suggested by @maisem

Also referred here

> ... would most likely require https://github.com/tailscale/tailscale/issues/465, or some workaround using subnet routers and externally assigned VIPs.
https://github.com/tailscale/tailscale/issues/502#issuecomment-1053766639


The proposal was to reserve any IP from the RFC1918 space (or really any space), and treat it as a VIP. This VIP is something that the ephemeral nodes would advertise using `--advertise-routes` and can be auto approved using [AutoApprovers]( https://tailscale.com/kb/1018/acls/#auto-approvers-for-routes-and-exit-nodes). You will still have to somehow make sure to listen on that VIP from your code though, and make your code dial out to the VIP instead of the Tailscale IP.

But as you mention, it's not user friendly :-)

**Comment by mayakacz** [comment]
> "any currently-online node with tag T". 

What does this represent? A specific service?

**Comment by josharian** [comment]
>> "any currently-online node with tag T".

> What does this represent? A specific service?

Yes, exactly. (Or slightly more precisely: A specific service in a specific environment, like prod or staging.)

**Comment by bradfitz** [comment]
Previously: https://github.com/tailscale/tailscale/issues/465 and https://github.com/tailscale/tailscale/issues/3860#issuecomment-1039240781

**Comment by bmizerany** [comment]
> > > "any currently-online node with tag T".
> 
> > What does this represent? A specific service?
> 
> Yes, exactly. (Or slightly more precisely: A specific service in a specific environment, like prod or staging.)

This would be extremely useful in proxy form, especially if shipped with Tailscale and was ready for config from the control plane if any should be configured.

@josharian Were you thinking of semantics along the lines of `"dashboard": ["tag:dashboard", "tag:prod"]` which translates to "All nodes that have **all of these tags** that this node's proxy can access/see?

**Comment by josharian** [comment]
I don't have a need for intersection of tags, but that sounds not unreasonable to me.

I do definitely still want this feature, though. :)

**Comment by damomurf** [comment]
Fwiw: this is what partially motivated me to implementing cname support in https://github.com/damomurf/coredns-tailscale (along with the vanity domain thing).

**Comment by patmaddox** [comment]
I think Tailscale is ideally suited to service discovery in this way.

Say I have two servers named sprl-prod-abc123 and sprl-prod-def456 both providing the sprl-prod service. It would be amazing if I could tag both machines with sprl-prod and then query sprl-prod.funky-animal.ts.net and get back the IP addresses.

Partitioning would happen by default as a result of ACLS, e.g. if nj-doh doesn’t have access to ny-doh, then when a nj-doh system DNS queries ny-doh-service.funky-animal.ts.net, it would get no results.

**Comment by patmaddox** [comment]
Related to this, I'd like to be able to run `tailscale cert` to get a certificate for any tag the host has (or possibly a single cert with subject alt names). That way I can script a cert generation on my tailscale host and config apps to use it.

**Comment by shubb30** [comment]
This is also needed in order to get high availability for Kubernetes proxied services, since the pods would be getting random hostnames based on the deployment name.

**Comment by jmkgreen** [comment]
I'm using AWS Spot instead of an On-Demand EC2. The new machine has a new private IP every time as expected. This ticket would let me reference a DNS name which is nice. Looking forward to it!

**Comment by Radiergummi** [comment]
Just to add another use-case here; we're running internal services in a Docker Swarm mode cluster, so we have enumerated nodes (node-01-swarm-01.funky-animal.ts.net etc.)  
The only good options to expose those services to our employees are using one of the cluster nodes as the entry point, or adding another "jump host" node to reverse proxy to the cluster nodes, either way artificially creating a single point of failure.

> Say I have two servers named sprl-prod-abc123 and sprl-prod-def456 both providing the sprl-prod service. It would be amazing if I could tag both machines with sprl-prod and then query sprl-prod.funky-animal.ts.net and get back the IP addresses.

The suggestion by @patmaddox sounds ideal. By just returning a bunch of A records for every node tagged with a given tag the user has access to, we get automatic service discovery the way it was meant to be, without any manual DNS shenanigans necessary.  
I'd love to see this implemented.

**Comment by damajor** [comment]
Naming this new feature "MagicTAGS" would be awesome 😄

**Comment by lf-** [comment]
The reason I really want this is to put a nix cache behind Tailscale at work and have it have good performance (it's highly latency and bandwidth sensitive). Currently it's in us-east-1 and the performance is terrible due to dubious peering of residential ISPs, compared to if it were on a CDN.

Putting it on a CDN directly isn't great because we don't have ways of authenticating clients easily without configuration, something which Tailscale simply fixes completely.

**Comment by lf-** [comment]
Hrm, this looks suspiciously like there is something that does what I want today. It's extremely exciting that *something* like this exists: https://tailscale.com/kb/1115/high-availability https://tailscale.com/kb/1281/app-connectors

**Comment by ScottJWalter** [comment]
Yes, but app connectors require a physical "connector service" on the tailnet, which usually means a separate box (like an rpi) or service somewhere in the network in front of the servers being balanced.  This doesn't solve HA for me (just moves the single point of failure to another point in the flow).

**Comment by kabirsikand** [comment]
We solved this in a slightly different way with a declarative config file and the serve command via a new Tailscale Services feature. Tailscale Services assigns a virtual tailscale ipv4/v6 address pair (TailVIP) and MagicDNS name and points to a set of service hosts; selection criteria includes Regional Routing. Read more here:  https://tailscale.com/blog/services-beta

**Comment by Radiergummi** [comment]
> We solved this in a slightly different way with a declarative config file and the serve command via a new Tailscale Services feature. Tailscale Services assigns a virtual tailscale ipv4/v6 address pair (TailVIP) and MagicDNS name and points to a set of service hosts; selection criteria includes Regional Routing. Read more here: https://tailscale.com/blog/services-beta

That still doesn't work for many use cases unfortunately, since it still requires a `tailscale serve` running somewhere; since it only (and arbitrarily so) supports to proxy to localhost, you cannot run it within a container, for example. So you're limited to exposing ports on a host…

**Comment by kabirsikand** [comment]
Thanks for all the feedback and input over the year(s)! This ticket is effectively closed with the release of Tailscale Services. 

If there are other use cases we'd love to hear them! We're likely to build on that platform for these types of requests. Addressing some of the questions above:
> it only (and arbitrarily so) supports to proxy to localhost

We now support remote destinations on the Tailscale Services platform. We're still working this into our documentation, so bear with us while we do so. 

> cannot run it within a container

We'd consider this a temporary limitation. We are aware of issues like https://github.com/tailscale/tailscale/issues/17769 and https://github.com/tailscale/tailscale/issues/18255 which are being actively prioritized. After that, Services should be usable within containerized environments.

**Comment by fitz-vivodyne** [comment]
@kabirsikand A couple tailscale services wishlist items:
1. automatic http -> https redirection
2. easier cert management. We're currently using tailscale services in containerized environments with ephemeral nodes. Biggest pain point at the moment is slow cert provisioning + hitting letsencrypt quotas. Working around that by creating an EFS volume to mount `/var/lib/tailscale/certs` to be shared across instances.

Edit: Loving services so far, very glad to see this in the wild
